File: templates/javaee-app/model/src/main/java/org/sgodden/tom/model/CustomerOrder.java
package org.sgodden.tom.model;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Configurable;

import javax.validation.constraints.NotNull;
import java.util.Collections;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * An order from a customer to transport goods from an origin
 * to a destination.
 * @author sgodden
 */
@SuppressWarnings("serial")
@Configurable
public class CustomerOrder extends AbstractIdentity implements ICustomerOrder {

    @NotNull
    private String customerReference;

    @NotNull
    private String orderNumber;

    @NotNull
    private CustomerOrderStatus status;

    private CollectionDetails collectionDetails;

    private DeliveryDetails deliveryDetails;

    private Set<CustomerOrderLine> orderLines = new HashSet<CustomerOrderLine>();

    @Autowired
    private Map<String, CustomerOrderState> stateObjects;

    public CustomerOrder() {
        status = CustomerOrderStatus.NEW;
    }

    public void cancel() {
        getStateObject().cancel(this);
    }

    public void confirm() {
        getStateObject().confirm(this);
    }

    public void ship() {
        getStateObject().ship(this);
    }

    public String getCustomerReference() {
        return customerReference;
    }

    public void setCustomerReference(String customerReference) {
        this.customerReference = customerReference;
    }

    /**
     * Returns the company's order number.
     * @return the company's order number.
     */
    public String getOrderNumber() {
        return orderNumber;
    }

    public void setOrderNumber(String orderNumber) {
        this.orderNumber = orderNumber;
    }

    public ICollectionDetails getCollectionDetails() {
        return collectionDetails;
    }

    public void setCollectionDetails(ICollectionDetails collectionDetails) {
        this.collectionDetails = (CollectionDetails) collectionDetails;
    }

    public IDeliveryDetails getDeliveryDetails() {
        return deliveryDetails;
    }

    public void setDeliveryDetails(IDeliveryDetails deliveryDetails) {
        this.deliveryDetails = (DeliveryDetails) deliveryDetails;
    }

    /**
     * Returns an immutable set of the lines on this order.
     * @return an immutable set of the lines on this order.
     */
    public Set<ICustomerOrderLine> getOrderLines() {
        Set<ICustomerOrderLine> ret = new HashSet<ICustomerOrderLine>();
        ret.addAll(orderLines);
        return Collections.unmodifiableSet(ret);
    }

    /**
     * Adds a line to this order.
     * @param line the line to add.
     */
    public void addOrderLine(ICustomerOrderLine line) {
        orderLines.add((CustomerOrderLine)line);
    }

    /**
     * Removes a line from this order.
     * @param line the line to remove.
     */
    public void removeOrderLine(ICustomerOrderLine line) {
        orderLines.remove(line);
    }

    public CustomerOrderStatus getStatus() {
        return status;
    }

    // intentionally package private
    void setStatus(CustomerOrderStatus status) {
        this.status = status;
    }

    private CustomerOrderState getStateObject() {
        return stateObjects.get(getStatus().name());
    }

    public Map<String, CustomerOrderState> getStateObjects() {
        return stateObjects;
    }

    public void setStateObjects(Map<String, CustomerOrderState> stateObjects) {
        this.stateObjects = stateObjects;
    }
}

File: templates/scala-mongodb/model/src/main/scala/org/sgodden/tom/model/CustomerOrder.scala
package org.sgodden.tom.model

import javax.validation.constraints.{Pattern,NotNull}
import collection.mutable.{HashSet, Set => MutableSet}
import org.springframework.beans.factory.annotation.{Configurable, Autowired}
import java.util.{Date, Map => JavaMap}
import org.joda.time.{DateTime}
import collection.JavaConverters
import JavaConverters._
import javax.validation.{ConstraintViolation, Validation}
import org.slf4j.{LoggerFactory, Logger}

// TODO - modify to use joda time
@Configurable
class CustomerOrder() extends ICustomerOrder {

  var _id: String = null

  @NotNull
  @Pattern(regexp = "cr.*", message = "{customerReferenceMustBeginWithCr}")
  var customerReference: String = null
  @NotNull
  var orderNumber: String = null
  @NotNull
  var status: CustomerOrderStatus.Value = CustomerOrderStatus.NEW
  @NotNull
  var bookingDate: DateTime = new DateTime(new Date().getTime)
  var collectionDetails: CollectionDetails = null
  var deliveryDetails: DeliveryDetails = null
  val orderLines: MutableSet[CustomerOrderLine] = new HashSet[CustomerOrderLine]

  @Autowired
  private var stateObjects: JavaMap[String, CustomerOrderState] = null

  def getId = _id

  def setId(id: String) {
    this._id = id
  }

  override def cancel = {
    getStateObject.cancel(this)
  }

  override def confirm = {
    getStateObject.confirm(this)
  }

  override def ship = {
    getStateObject.ship(this)
  }

  def approveRemove(onApprove: () => Unit) = {
    onApprove()
  }

  def approvePersist(onApprove: () => Unit) = {
    validate
    onApprove()
  }

  def validate = {
    val constraints = CustomerOrder.validator.validate(this).asScala.toSet
    if (constraints.size > 0) {
      if (CustomerOrder.LOG.isDebugEnabled) {
        for (violation <- constraints) {
          CustomerOrder.LOG.debug(violation.toString)
        }
      }
      throw new ValidationException(constraints.asInstanceOf[Set[ConstraintViolation[AnyRef]]])
    }

  }

  override def getBookingDate = bookingDate
  override def setBookingDate(date: DateTime) = this.bookingDate = date

  override def getCustomerReference = customerReference
  override def setCustomerReference(customerReference: String) = this.customerReference = customerReference

  override def getOrderNumber: String = orderNumber
  override def setOrderNumber(orderNumber: String) = this.orderNumber = orderNumber

  override def getCollectionDetails: ICollectionDetails = collectionDetails
  override def setCollectionDetails(collectionDetails: ICollectionDetails) =
    this.collectionDetails = collectionDetails.asInstanceOf[CollectionDetails]

  override def getDeliveryDetails: IDeliveryDetails = deliveryDetails
  override def setDeliveryDetails(deliveryDetails: IDeliveryDetails) =
    this.deliveryDetails = deliveryDetails.asInstanceOf[DeliveryDetails]

  override def getOrderLines: Set[ICustomerOrderLine] = orderLines.map(_.asInstanceOf[ICustomerOrderLine]).toSet
  override def addOrderLine(line: ICustomerOrderLine) =
    orderLines.add(line.asInstanceOf[CustomerOrderLine])
  override def removeOrderLine(line: ICustomerOrderLine) =
    orderLines.remove(line.asInstanceOf[CustomerOrderLine])

  override def getStatus: CustomerOrderStatus.Value = status
  def setStatus(status: CustomerOrderStatus.Value) = this.status = status

  private def getStateObject: CustomerOrderState = stateObjects.get(getStatus.toString)

  def getStateObjects: JavaMap[String, CustomerOrderState] = stateObjects
  def setStateObjects(stateObjects: JavaMap[String, CustomerOrderState]) = this.stateObjects = stateObjects
}

object CustomerOrder {
  private final val LOG: Logger = LoggerFactory.getLogger(classOf[CustomerOrder])
  val validator = Validation.buildDefaultValidatorFactory.getValidator
}
